cmd/process-user-data/provision.go

```diff
@@ -5,166 +5,189 @@ import (
+	return &cc, nil
+}
 
-	// Create the directory.
-	err := os.MkdirAll(dir, 0755)
+func parseDaemonConfig(content []byte) (*daemon.Config, error) {
+	var dc daemon.Config
+	err := json.Unmarshal(content, &dc)
 	if err != nil {
-		return fmt.Errorf("failed to create directory: %s", err)
+		return nil, err
+	}
+	return &dc, nil
+}
 
+func getDaemonConfigContent(cc *CloudConfig) ([]byte, error) {
+	for _, wf := range cc.WriteFiles {
+		if wf.Path != cfg.daemonConfigPath {
+			continue
+		}
+		return []byte(wf.Content), nil
 	}
+	return nil, fmt.Errorf("failed to find entry for %s in cloud config", cfg.daemonConfigPath)
+}
 
+func writeAuthJson(dc *daemon.Config) error {
 	// Create the file
-	file, err := os.Create(dstFilePath)
+	file, err := os.Create(cfg.authJsonPath)
 	if err != nil {
-		return fmt.Errorf("failed to create file: %s", err)
+		return fmt.Errorf("failed to create file: %w", err)
 	}
 	defer file.Close()
 
-	// Write userData to file
-	_, err = file.WriteString(userData)
+	// Write the authJson to the file
+	_, err = file.WriteString(dc.AuthJson)
 	if err != nil {
-		return fmt.Errorf("failed to write userData to file: %s", err)
+		return fmt.Errorf("failed to write authJson to file: %w", err)
 	}
-
-	fmt.Printf("Wrote userData to file: %s\n", dstFilePath)
-
 	return nil
-
 }
 
-// Get daemon.Config from userData
-func getConfigFromUserData(userData string) daemon.Config {
-
-	// UnMarshal the userData into forwarder (daemon) Config struct
-	var daemonConfig daemon.Config
-
-	err := json.Unmarshal([]byte(userData), &daemonConfig)
+func writeDaemonConfig(content []byte) error {
+	err := os.WriteFile(cfg.daemonConfigPath, content, 0644)
 	if err != nil {
-		fmt.Printf("failed to unmarshal userData: %s\n", err)
-		return daemon.Config{}
+		return fmt.Errorf("failed to write daemon config file: %w", err)
 	}
+	fmt.Printf("Wrote daemon config file: %s\n", cfg.daemonConfigPath)
+	return nil
+}
 
-	return daemonConfig
+type UserDataProvider interface {
+	GetUserData(ctx context.Context) ([]byte, error)
+	GetRetryDelay() time.Duration
 }
 
-func provisionFiles(cmd *cobra.Command, args []string) error {
+type DefaultRetry struct{}
 
-	var (
-		ctx    context.Context
-		cancel context.CancelFunc
-	)
+func (d DefaultRetry) GetRetryDelay() time.Duration {
+	return 5 * time.Second
+}
 
-	// Use retry.Do to retry the getUserData function until it succeeds
-	// This is needed because the VM's userData is not available immediately
-	// Have an option to either wait forever or timeout after a certain amount of time
-	// https://github.com/avast/retry-go
+type AzureUserDataProvider struct{ DefaultRetry }
 
-	// Create context with the specified timeout
-	ctx, cancel = context.WithTimeout(context.Background(), time.Duration(cfg.userDataFetchTimeout)*time.Second)
-	defer cancel()
+func (a AzureUserDataProvider) GetUserData(ctx context.Context) ([]byte, error) {
+	url := azure.AzureUserDataImdsUrl
+	fmt.Printf("provider: Azure, userDataUrl: %s\n", url)
```

Comment:
All the `fmt.Printf(....\n")` can be replaced with `fmt.Println`.

Reply:
Ideally we should be using some sort of logging library than printfs because that gives more metadata context and structure.

Reply:
I agree, but this would exceed the scope of the PRs. I have a follow-up PR scheduled which aligns the code organization of the tool with the other golang code in this repo, we can then switch to a common logging method.

Reply:
Yep I agree, I just made a general comment.